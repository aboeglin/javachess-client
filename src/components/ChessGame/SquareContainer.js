import React from "react";
import { PropTypes } from "prop-types";
import { curry, find, pipe, propOr } from "ramda";

import Piece from "./Piece";
import { Square, SquareLabel } from "./styled";
import withChess from "../../hoc/withChess";

const getPieceAtSquare = curry((pieces, square) =>
  find(({ x, y }) => x === square.x && y === square.y)(pieces)
);

const isSelected = curry(
  (selection, square) =>
    square.x === propOr(false, "x", selection) &&
    square.y === propOr(false, "y", selection)
);

const getSelectionColor = selection => (selection ? selection.color : false);

export const SquareContainer = ({ square, pieces, pieceMoved, selection }) =>
  pipe(getPieceAtSquare(pieces), piece => (
    <Square
      color={square.color}
      x={square.x}
      y={square.y}
      key={`${square.x}${square.y}`}
      highlighted={square.highlighted}
      selected={isSelected(selection, square)}
      selectionColor={getSelectionColor(selection)}
      onClick={() => {
        if (
          selection &&
          (selection.x !== square.x || selection.y !== square.y)
        ) {
          if (!piece || piece.color !== selection.color) {
            pieceMoved(selection.x, selection.y, square.x, square.y);
          }
        }
      }}
    >
      <SquareLabel color={square.color}>
        {square.x}
        {square.y}
      </SquareLabel>
      {piece && <Piece piece={piece} />}
    </Square>
  ))(square);

SquareContainer.propTypes = {
  square: PropTypes.object,
  pieces: PropTypes.array
};

export default withChess(SquareContainer);
