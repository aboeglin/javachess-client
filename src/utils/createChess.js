import SockJS from "sockjs-client";
import Stomp from "stompjs";

import { pipe, prop, append, forEach } from "ramda";

const HARDCODED_EMAIL = "user@domain.tld";

const createChess = () => {
  const socket = new SockJS("http://localhost:8080/ws");
  const ws = Stomp.over(socket);

  ws.debug = null;

  let observers = [];
  let gameId = null;

  let state = {
    pieces: [],
    possibleMoves: [],
    selection: null
  };

  ws.connect({}, () => {
    ws.subscribe("/user/queue/lfg/ack", frame => {
      gameId = pipe(prop("body"), JSON.parse, prop("gameId"))(frame);
      ws.subscribe(
        `/queue/game/${gameId}/ready`,
        pipe(prop("body"), body => {
          const newPieces = JSON.parse(body).game.board.pieces;
          updateState({ pieces: newPieces });
        })
      );

      ws.subscribe(
        `/queue/game/${gameId}/piece-moved`,
        pipe(prop("body"), body => {
          const newPieces = JSON.parse(body).game.board.pieces;

          // If move successful
          updateState({
            pieces: newPieces,
            possibleMoves: [],
            selection: null
          });
          // Else API should return MoveError and we should keep the possible moves
        })
      );

      ws.subscribe(
        `/queue/game/${gameId}/possible-moves`,
        pipe(prop("body"), body => {
          const possibleMoves = JSON.parse(body).possibleMoves;
          updateState({ possibleMoves: possibleMoves });
        })
      );

      ws.send(
        `/app/game/${gameId}/join`,
        {},
        JSON.stringify({ email: HARDCODED_EMAIL })
      );
    });

    ws.subscribe("/user/queue/errors", frame => {
      console.error(frame.body);
    });

    ws.send("/app/lfg", {}, JSON.stringify({ email: HARDCODED_EMAIL }));
  });

  const pieceSelected = piece => {
    updateState({ selection: piece });
    ws.send(
      `/app/game/${gameId}/select-piece`,
      {},
      JSON.stringify({ email: HARDCODED_EMAIL, x: piece.x, y: piece.y })
    );
  };

  const pieceMoved = (fromX, fromY, toX, toY) => {
    ws.send(
      `/app/game/${gameId}/perform-move`,
      {},
      JSON.stringify({ email: HARDCODED_EMAIL, fromX, fromY, toX, toY })
    );
  };

  const updateState = newState => {
    state = { ...state, ...newState };
    forEach(fn => fn(state))(observers);
  };

  // Store array of handlers, otherwise only the last one can get notified from state changed
  const onStateChanged = fn => {
    observers = append(fn, observers);
    fn(state);
  };

  return {
    onStateChanged,
    pieceSelected,
    pieceMoved
  };
};

export default createChess;
