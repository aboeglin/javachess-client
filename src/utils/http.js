import F from "fluture";
import axios from "axios";
import { curry } from "ramda";

export const post = curry((url, body) =>
  F((bad, good) => {
    axios
      .post(url, body)
      .then(good)
      .catch(bad);

    // Cancelation function
    return () => {};
  })
);

export const patch = curry((url, body) =>
  F((bad, good) => {
    axios
      .patch(url, body)
      .then(good)
      .catch(bad);

    // Cancelation function
    return () => {};
  })
);

export const get = (url) =>
  F((bad, good) => {
    axios
      .get(url)
      .then(good)
      .catch(bad);

    // Cancelation function
    return () => {};
  });
