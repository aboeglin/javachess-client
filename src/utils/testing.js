import React from "react";
import ReactDOM from "react-dom";
import { keys, pipe, propOr, curry } from "ramda";
import renderer from "react-test-renderer";
import cleanStack from "clean-stack";
import { Helmet } from "react-helmet";

beforeEach(() => {
  process.env.GATSBY_ENTRYPOINT = "/";
});

export const hasNonEmptyPropTypes = x => {
  const propTypes = propOr(false, "propTypes", x);
  // () => <span>cool</span> is valid, so no need for propTypes
  if (!propTypes) return true;
  // expect some keys if defined
  return pipe(keys, z => z.length > 0)(propTypes);
};

export const render = (C, props = {}, el = "div") => {
  try {
    const div = document.createElement(el);
    ReactDOM.render(
      <>
        <div id="leafletmap" />
        <C {...props} />
      </>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  } catch (error) {
    console.error(error.stack ? cleanStack(error.stack) : error);
  }
};
export const renderize = (C, props = {}, el = "div") => () => {
  try {
    const rend = () => {
      try {
        render(C, props, el);
      } catch (error) {
        if (error.stack) error.stack = cleanStack(error.stack);
        console.error(error.stack ? error.stack : error);
        throw error;
      }
    };
    expect(rend).not.toThrow();
  } catch (error) {
    console.error(error.stack ? cleanStack(error.stack) : error);
  }
};

const nameOf = C => C.displayName || C.name || "A Component";

export const renderWithTestRenderer = curry((C, props) => {
  it(`renders <${nameOf(C)} />`, () => {
    try {
      const c = renderer.create(<C {...props} />);
      let tree = c.toJSON();
      expect(tree).toMatchSnapshot();
    } catch (error) {
      console.log(error.stack ? cleanStack(error.stack) : error);
    }
  });
});

export const standardComponentTest = (C, props = {}, el = "div") => {
  renderWithTestRenderer(C, props);
  it("renders without crashing", renderize(C, props, el));
  it("has propTypes", () => {
    expect(hasNonEmptyPropTypes(C)).toBeTruthy();
  });
};

export const standardComponentTestWithHeadTest = (
  C,
  props = {},
  el = "div",
  headProperty = "",
  headData = null
) => {
  renderWithTestRenderer(C, props);
  it("renders without crashing", renderize(C, props, el));
  it("has propTypes", () => {
    expect(hasNonEmptyPropTypes(C)).toBeTruthy();
  });
  it("has correct head props", () => {
    headData
      ? expect(Helmet.peek()).toHaveProperty(headProperty, headData)
      : expect(Helmet.peek()).toHaveProperty(headProperty);
  });
};
