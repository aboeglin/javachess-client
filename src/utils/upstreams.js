import { __, concat, curry, map, pipe, prop, toString } from "ramda";

import { get, patch, post } from "./http";

const ENDPOINT =
  process.env.GATSBY_CHESS_SERVICE_ENDPOINT || "http://localhost:8080";
const GET_GAMES_PATH = "/games";
const POST_GAMES_PATH = "/games";
const PATCH_GAMES_PATH = "/games";

export const fetchGamesWithGetAndEndpoint = curry((get, endpoint, path) =>
  pipe(concat(endpoint), get, map(prop("data")))(path)
);

export const fetchGames = () =>
  fetchGamesWithGetAndEndpoint(get, ENDPOINT, GET_GAMES_PATH);

export const createGameWithPostAndEndpoint = curry(
  (post, endpoint, path, userName) =>
    pipe(
      concat(endpoint),
      post(__, { userId: userName }),
      map(prop("data"))
    )(path)
);

export const createGame = createGameWithPostAndEndpoint(
  post,
  ENDPOINT,
  POST_GAMES_PATH
);

export const joinGameWithPatchAndEndpoint = curry(
  (patch, endpoint, path, userName, gameId) =>
    pipe(
			toString,
      concat("/"),
      concat(path),
      concat(endpoint),
      patch(__, { userId: userName }),
      map(prop("data"))
    )(gameId)
);

export const joinGame = joinGameWithPatchAndEndpoint(patch, ENDPOINT, PATCH_GAMES_PATH);
