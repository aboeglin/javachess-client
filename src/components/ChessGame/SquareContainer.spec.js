import { SquareContainer, handleClick } from "./SquareContainer";
import { standardComponentTest } from "../../utils/testing";

standardComponentTest(SquareContainer, {
  pieces: [],
  square: { color: "WHITE", x: "b", y: "3", highlighted: false }
});

standardComponentTest(SquareContainer, {
  selection: { x: "a", y: "2", color: "WHITE" },
  pieces: [{ x: "a", y: "3", type: "PAWN", color: "WHITE" }],
  square: { color: "WHITE", x: "a", y: "3", highlighted: false }
});

test("handleClick", done => {
  const square = {
    x: "b",
    y: "4"
  };

  const selection = {
    x: "b",
    y: "2",
    color: "WHITE"
  };

  const piece = null;

  const pieceMoved = (...args) => {
    expect(args).toEqual(["b", "2", "b", "4"]);
    done();
  };

  const event = {};

  handleClick(pieceMoved, selection, piece, square)(event);
});

test("handleClick should not be called when there is no selection", () => {
  const square = {
    x: "b",
    y: "4"
  };

  const selection = null;

  const piece = null;

  const pieceMoved = jest.fn();

  const event = {};
  handleClick(pieceMoved, selection, piece, square)(event);
  expect(pieceMoved).not.toHaveBeenCalled();
});

test("handleClick should not be called when the selection and the piece at destination have the same color", () => {
  const square = {
    x: "b",
    y: "4"
  };

  const selection = {
    x: "b",
    y: "2",
    color: "WHITE"
  };

  const piece = {
    color: "WHITE"
  };

  const pieceMoved = jest.fn();

  const event = {};
  handleClick(pieceMoved, selection, piece, square)(event);
  expect(pieceMoved).not.toHaveBeenCalled();
});

test("handleClick should not be called if the square is where the selected piece is", () => {
  const square = {
    x: "b",
    y: "2"
  };

  const selection = {
    x: "b",
    y: "2",
    color: "WHITE"
  };

  const piece = {
    x: "b",
    y: "2",
    color: "WHITE"
  };

  const pieceMoved = jest.fn();

  const event = {};
  handleClick(pieceMoved, selection, piece, square)(event);
  expect(pieceMoved).not.toHaveBeenCalled();
});
