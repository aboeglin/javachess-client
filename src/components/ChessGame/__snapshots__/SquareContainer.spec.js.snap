// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders <SquareContainer /> 1`] = `
<div
  className="css-14mmxv0"
  color="WHITE"
  onClick={[Function]}
  selected={false}
  x="b"
  y="3"
>
  <div
    className="css-1qigpt8"
    color="WHITE"
  >
    b
    3
  </div>
</div>
`;

exports[`renders <SquareContainer /> 2`] = `
<div
  className="css-14mmxv0"
  color="WHITE"
  onClick={[Function]}
  selected={false}
  x="a"
  y="3"
>
  <div
    className="css-1qigpt8"
    color="WHITE"
  >
    a
    3
  </div>
  <div
    className="css-kxoee0"
    onClick={[Function]}
    type="PAWN"
    x="a"
  >
    <img
      src={Object {}}
    />
  </div>
</div>
`;
